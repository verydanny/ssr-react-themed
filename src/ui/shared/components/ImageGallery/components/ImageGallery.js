import React, { Component } from 'react'
import { Photo } from '@rentpath/react-ui-core'
import PropTypes from 'prop-types'
import { themed } from 'react-themed-too'
import Slider from 'react-image-gallery'
import LazyLoad from 'react-lazyload'
import autobind from 'autobind-decorator'
import getOr from 'lodash/fp/getOr'
import GalleryNav from './GalleryNav'

const SWIPE_THRESHOLD = 20
const FLICK_THRESHOLD = 50
const SLIDE_DURATION = 300

@themed('*')
export default class ImageGallery extends Component {
  static propTypes = {
    photos: PropTypes.array,
    theme: PropTypes.object,
    viewType: PropTypes.string,
    className: PropTypes.string,
    dataTagItem: PropTypes.string,
    dataTagSection: PropTypes.string,
    showIndex: PropTypes.bool,
    disableArrowKeys: PropTypes.bool,
    lazyload: PropTypes.bool,
    disableSwipe: PropTypes.bool,
    formattedBaseUrl: PropTypes.object,
    onIndexChange: PropTypes.func,
    onClick: PropTypes.func,
    startIndex: PropTypes.number,
  }

  static defaultProps = {
    photos: [],
    theme: {},
    viewType: 'map',
    lazyload: true,
    disableSwipe: true,
    formattedBaseUrl: {},
    startIndex: 0,
  }

  constructor(props) {
    super(props)
    this.firstPhotoPath = getOr(null, `photos[${props.startIndex}].path`, this.props)
    this.nextSlideIndex = 0
    this.previousIndex = null
    this.updatedFromTouch = false
    this.state = {
      renderOnlyFirstPhoto: true,
    }
  }

  /**
   * To support a different tagging call for swipe vs click,
   * we need a way to tell when the user swiped.
   * In addition, we need to determine if the user swiped left or right.
   * This method, in conjunction with handleOnSlide() lets us
   * determine all that.
   */
  @autobind
  handleTouchStart() {
    const {
      disableSwipe,
    } = this.props

    if (disableSwipe) return

    // Set a flag so we can check later and know that the user was swiping
    this.updatedFromTouch = true

    // Save the slide index so we can check later and determine
    // if the user swiped left or right
    this.previousIndex = this.imageGallery.getCurrentIndex()
  }

  @autobind
  disablePlaceholder() {
    // on first slide, disable placeholder
    if (this.state.renderOnlyFirstPhoto) {
      this.setState({ renderOnlyFirstPhoto: false })
    }
  }

  @autobind
  handleOnSlide(index) {
    const { dataTagSection, disableSwipe, onIndexChange } = this.props

    this.disablePlaceholder()

    // Expose change in index to parent
    if (onIndexChange) onIndexChange(index)

    if (disableSwipe) return

    // If the user didn't swipe, do nothing
    if (!this.updatedFromTouch || !window.eventTracker) return

    // Save the current slide index, so we can use it later
    // to calculate if the user swiped left vs right
    this.currentIndex = index

    // Compare the current slide to the previous slide
    // from when the user started swiping, to determine if the user
    // swiped left or right. Keep in mind the gallery can loop around
    // in either direction.
    const nextSlide = (
      (this.currentIndex === this.previousIndex + 1) ||
      (this.currentIndex === 0 && this.previousIndex > 1)
    )
    const direction = nextSlide === true ? 'right' : 'left'

    window.eventTracker.track('swipe', {
      item: `${direction}_arrow`,
      action: 'swipe',
      section: dataTagSection,
    })

    // Clear the flag when the swipe action is done
    this.updatedFromTouch = false
  }

  arrowNavigation(onClick, disabled, direction = 'next') {
    return (
      <GalleryNav
        direction={direction}
        viewType={this.props.viewType}
        onClick={onClick}
        disabled={disabled}
      />
    )
  }

  @autobind
  renderItem(photo) {
    const {
      theme,
      viewType,
      formattedBaseUrl,
    } = this.props

    const {
      renderOnlyFirstPhoto,
    } = this.state

    const { path, caption } = photo
    const renderPlaceholder = renderOnlyFirstPhoto && path !== this.firstPhotoPath

    // on initial page load, render placeholder for second slide
    if (renderPlaceholder) return <div />

    const { jpg, jpgRetina, webp } = formattedBaseUrl
    // this assumes an API response of format `imgr/HASH/`
    const photoHash = path.split('/')[1]

    const sources = [
      {
        type: 'image/webp',
        srcset: `${webp}${photoHash}`,
      },
      {
        srcset: `${jpg}${photoHash}, ${jpgRetina}${photoHash} 2x`,
      },
    ]

    return (
      <Photo
        sources={sources}
        className={theme[`SliderImage-${viewType}`]}
        alt={caption}
        url={`${jpg}${photoHash}`}
      />
    )
  }

  @autobind
  renderLeftNav(onClick, disabled) {
    return this.arrowNavigation(onClick, disabled, 'prev')
  }

  @autobind
  renderRightNav(onClick, disabled) {
    return this.arrowNavigation(onClick, disabled)
  }

  render() {
    const {
      photos,
      theme,
      viewType,
      className,
      dataTagItem,
      showIndex,
      disableArrowKeys,
      lazyload,
      disableSwipe,
      dataTagSection,
      onClick,
      startIndex,
    } = this.props

    // Different props to use depending if you want swiping
    const swipeProps = disableSwipe ? {
      disableSwipe: true,
      preventDefaultTouchmoveEvent: true,
    } : {
      swipeThreshold: SWIPE_THRESHOLD,
      flickThreshold: FLICK_THRESHOLD,
      slideDuration: SLIDE_DURATION,
      onTouchStart: this.handleTouchStart,
    }

    if (photos && photos.length) {
      const slider = (
        <Slider
          ref={comp => { this.imageGallery = comp }}
          lazyLoad
          showThumbnails={false}
          showPlayButton={false}
          showFullscreenButton={false}
          renderItem={this.renderItem}
          renderRightNav={this.renderRightNav}
          renderLeftNav={this.renderLeftNav}
          items={photos}
          showIndex={photos.length > 1 ? showIndex : false}
          disableArrowKeys={disableArrowKeys}
          onSlide={this.handleOnSlide}
          onClick={onClick}
          startIndex={startIndex}
          {...swipeProps}
        />
      )

      let gallery = slider

      if (lazyload) {
        gallery = (
          <LazyLoad
            height={158}
            width={280}
            resize
            offset={[200, -100]}
            overflow
          >
            {slider}
          </LazyLoad>
        )
      }

      return (
        <div
          className={className}
          data-tid="listing-photo-gallery"
          data-tag_item={dataTagItem}
          data-tag_section={dataTagSection}
        >
          {gallery}
        </div>
      )
    }
    return (
      <div className={theme[`SliderImage-${viewType}`]} />
    )
  }
}
