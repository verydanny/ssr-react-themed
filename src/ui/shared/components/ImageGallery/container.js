import { connect } from 'react-redux'
import { formattedBaseUrl } from 'app/selectors/listingPhotoFormats'
import { getCurrentView } from 'app/selectors/page'
import ImageGallery from './components/ImageGallery'

const mapStateToProps = state => ({
  formattedBaseUrl: formattedBaseUrl(state),
  viewType: getCurrentView(state),
})

const mapDispatchToProps = {}

export default connect(mapStateToProps, mapDispatchToProps)(ImageGallery)
