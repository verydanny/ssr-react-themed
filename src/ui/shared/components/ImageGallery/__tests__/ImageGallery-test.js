import React from 'react'
import { mount, shallow } from 'enzyme'

import ImageGallery from '../components/ImageGallery'

const ImageGalleryUnwrapped = ImageGallery.WrappedComponent

describe('ImageGallery', () => {
  const props = {
    formattedBaseUrl: {
      jpg: 'jpg/',
      jpgRetina: 'jpgRetina/',
      webp: 'webp/',
    },
    lazyload: true,
    photos: [
      { path: 'imgr/foo/', caption: 'bar' },
      { path: 'imgr/bam/', caption: 'baz' },
    ],
  }

  const nolazyloadProps = {
    formattedBaseUrl: {
      jpg: 'jpg/',
      jpgRetina: 'jpgRetina/',
      webp: 'webp/',
    },
    lazyload: false,
    photos: [
      { path: 'imgr/foo/', caption: 'bar' },
      { path: 'imgr/bam/', caption: 'baz' },
    ],
  }

  it('renders the gallery', () => {
    const wrapper = mount(<ImageGalleryUnwrapped {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders the gallery without Lazyload', () => {
    const wrapper = mount(<ImageGallery {...nolazyloadProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a LazyLoad', () => {
    const wrapper = mount(<ImageGallery {...props} />)
    const expected = 1
    const actual = wrapper.find('LazyLoad').length
    expect(actual).toEqual(expected)
  })

  it('renders only the first photo', () => {
    const expected = 1
    const wrapper = mount(<ImageGallery {...nolazyloadProps} />)
    const actual = wrapper.find('img').length
    expect(actual).toEqual(expected)
  })

  describe('ImageGallery photoRender', () => {
    const photo = props.photos[0]
    const wrapper = shallow(<ImageGalleryUnwrapped {...props} />)

    it('renders images', () => {
      expect(wrapper.instance().renderItem(photo)).toMatchSnapshot()
    })
  })
})
