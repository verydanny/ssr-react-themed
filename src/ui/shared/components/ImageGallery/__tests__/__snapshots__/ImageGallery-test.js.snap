// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery ImageGallery photoRender renders images 1`] = `
<Themed(Photo)
  alt="bar"
  sources={
    Array [
      Object {
        "srcset": "webp/foo",
        "type": "image/webp",
      },
      Object {
        "srcset": "jpg/foo, jpgRetina/foo 2x",
      },
    ]
  }
  url="jpg/foo"
/>
`;

exports[`ImageGallery renders the gallery 1`] = `
<ImageGallery
  disableSwipe={true}
  formattedBaseUrl={
    Object {
      "jpg": "jpg/",
      "jpgRetina": "jpgRetina/",
      "webp": "webp/",
    }
  }
  lazyload={true}
  photos={
    Array [
      Object {
        "caption": "bar",
        "path": "imgr/foo/",
      },
      Object {
        "caption": "baz",
        "path": "imgr/bam/",
      },
    ]
  }
  startIndex={0}
  theme={Object {}}
  viewType="map"
>
  <div
    data-tid="listing-photo-gallery"
  >
    <LazyLoad
      height={158}
      offset={
        Array [
          200,
          -100,
        ]
      }
      once={false}
      overflow={true}
      resize={true}
      scroll={true}
      unmountIfInvisible={false}
      width={280}
    >
      <div
        className="lazyload-placeholder"
        style={
          Object {
            "height": 158,
          }
        }
      />
    </LazyLoad>
  </div>
</ImageGallery>
`;

exports[`ImageGallery renders the gallery without Lazyload 1`] = `
<Themed(ImageGallery)
  formattedBaseUrl={
    Object {
      "jpg": "jpg/",
      "jpgRetina": "jpgRetina/",
      "webp": "webp/",
    }
  }
  lazyload={false}
  photos={
    Array [
      Object {
        "caption": "bar",
        "path": "imgr/foo/",
      },
      Object {
        "caption": "baz",
        "path": "imgr/bam/",
      },
    ]
  }
>
  <ImageGallery
    disableSwipe={true}
    formattedBaseUrl={
      Object {
        "jpg": "jpg/",
        "jpgRetina": "jpgRetina/",
        "webp": "webp/",
      }
    }
    lazyload={false}
    photos={
      Array [
        Object {
          "caption": "bar",
          "path": "imgr/foo/",
        },
        Object {
          "caption": "baz",
          "path": "imgr/bam/",
        },
      ]
    }
    startIndex={0}
    theme={Object {}}
    viewType="map"
  >
    <div
      data-tid="listing-photo-gallery"
    >
      <ImageGallery
        autoPlay={false}
        disableArrowKeys={false}
        disableSwipe={true}
        disableThumbnailScroll={false}
        flickThreshold={0.4}
        indexSeparator=" / "
        infinite={true}
        items={
          Array [
            Object {
              "caption": "bar",
              "path": "imgr/foo/",
            },
            Object {
              "caption": "baz",
              "path": "imgr/bam/",
            },
          ]
        }
        lazyLoad={true}
        onSlide={[Function]}
        preventDefaultTouchmoveEvent={true}
        renderFullscreenButton={[Function]}
        renderItem={[Function]}
        renderLeftNav={[Function]}
        renderPlayPauseButton={[Function]}
        renderRightNav={[Function]}
        showBullets={false}
        showFullscreenButton={false}
        showIndex={false}
        showNav={true}
        showPlayButton={false}
        showThumbnails={false}
        slideDuration={450}
        slideInterval={3000}
        startIndex={0}
        stopPropagation={false}
        swipeThreshold={30}
        swipingTransitionDuration={0}
        thumbnailPosition="bottom"
        useBrowserFullscreen={true}
        useTranslate3D={true}
      >
        <div
          aria-live="polite"
          className="image-gallery "
        >
          <div
            className="image-gallery-content"
          >
            <div
              className="image-gallery-slide-wrapper bottom"
            >
              <span
                key="navigation"
              >
                <Themed(GalleryNav)
                  direction="prev"
                  disabled={false}
                  onClick={[Function]}
                  viewType="map"
                >
                  <GalleryNav
                    direction="prev"
                    disabled={false}
                    onClick={[Function]}
                    theme={Object {}}
                    viewType="map"
                  >
                    <div
                      className=""
                    >
                      <span
                        className=""
                        data-nopdplink={true}
                        data-tag_item="left_arrow"
                        onClick={[Function]}
                        role="button"
                        tabIndex={0}
                      >
                        <Themed(Icon)
                          data-nopdplink={true}
                          name="chevronLeft"
                          svgs={
                            Object {
                              "chevronLeft": "",
                              "chevronRight": "",
                            }
                          }
                        >
                          <Icon
                            data-nopdplink={true}
                            name="chevronLeft"
                            svgs={
                              Object {
                                "chevronLeft": "",
                                "chevronRight": "",
                              }
                            }
                            theme={Object {}}
                          />
                        </Themed(Icon)>
                      </span>
                    </div>
                  </GalleryNav>
                </Themed(GalleryNav)>
                <Themed(GalleryNav)
                  direction="next"
                  disabled={false}
                  onClick={[Function]}
                  viewType="map"
                >
                  <GalleryNav
                    direction="next"
                    disabled={false}
                    onClick={[Function]}
                    theme={Object {}}
                    viewType="map"
                  >
                    <div
                      className=""
                    >
                      <span
                        className=""
                        data-nopdplink={true}
                        data-tag_item="right_arrow"
                        onClick={[Function]}
                        role="button"
                        tabIndex={0}
                      >
                        <Themed(Icon)
                          data-nopdplink={true}
                          name="chevronRight"
                          svgs={
                            Object {
                              "chevronLeft": "",
                              "chevronRight": "",
                            }
                          }
                        >
                          <Icon
                            data-nopdplink={true}
                            name="chevronRight"
                            svgs={
                              Object {
                                "chevronLeft": "",
                                "chevronRight": "",
                              }
                            }
                            theme={Object {}}
                          />
                        </Themed(Icon)>
                      </span>
                    </div>
                  </GalleryNav>
                </Themed(GalleryNav)>
              </span>
              <Swipeable
                className="image-gallery-swipe"
                delta={0}
                disabled={true}
                flickThreshold={0.4}
                key="swipeable"
                nodeName="div"
                onSwiped={[Function]}
                onSwiping={[Function]}
                preventDefaultTouchmoveEvent={true}
                stopPropagation={false}
              >
                <div
                  className="image-gallery-swipe"
                >
                  <div
                    className="image-gallery-slides"
                  >
                    <div
                      className="image-gallery-slide center"
                      key="0"
                      style={
                        Object {
                          "MozTransform": "translate3d(0%, 0, 0)",
                          "OTransform": "translate3d(0%, 0, 0)",
                          "WebkitTransform": "translate3d(0%, 0, 0)",
                          "msTransform": "translate3d(0%, 0, 0)",
                          "transform": "translate3d(0%, 0, 0)",
                        }
                      }
                    >
                      <Themed(Photo)
                        alt="bar"
                        sources={
                          Array [
                            Object {
                              "srcset": "webp/foo",
                              "type": "image/webp",
                            },
                            Object {
                              "srcset": "jpg/foo, jpgRetina/foo 2x",
                            },
                          ]
                        }
                        url="jpg/foo"
                      >
                        <Photo
                          alt="bar"
                          sources={
                            Array [
                              Object {
                                "srcset": "webp/foo",
                                "type": "image/webp",
                              },
                              Object {
                                "srcset": "jpg/foo, jpgRetina/foo 2x",
                              },
                            ]
                          }
                          theme={
                            Object {
                              "Photo": undefined,
                            }
                          }
                          url="jpg/foo"
                        >
                          <picture>
                            <source
                              key="0webp/foo"
                              srcSet="webp/foo"
                              type="image/webp"
                            />
                            <source
                              key="1jpg/foo, jpgRetina/foo 2x"
                              srcSet="jpg/foo, jpgRetina/foo 2x"
                            />
                            <img
                              alt="bar"
                              className=""
                              data-tid="photo"
                              onError={[Function]}
                              src="jpg/foo"
                            />
                          </picture>
                        </Photo>
                      </Themed(Photo)>
                    </div>
                    <div
                      className="image-gallery-slide right"
                      key="1"
                      style={
                        Object {
                          "MozTransform": "translate3d(100%, 0, 0)",
                          "OTransform": "translate3d(100%, 0, 0)",
                          "WebkitTransform": "translate3d(100%, 0, 0)",
                          "msTransform": "translate3d(100%, 0, 0)",
                          "transform": "translate3d(100%, 0, 0)",
                        }
                      }
                    >
                      <div />
                    </div>
                  </div>
                </div>
              </Swipeable>
            </div>
          </div>
        </div>
      </ImageGallery>
    </div>
  </ImageGallery>
</Themed(ImageGallery)>
`;
